### PR TITLE
Adding media to feeds

### PR DESCRIPTION
**Pull request by SaboohTaha**
Added option to add media (images/videos) while creating new feed
On edit page you can add or remove media files
Created a slider to view multiple media in a feed

### CODE REVIEW

File: src/components/Feed/Edit.jsx [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+        for (var key in feed) {
+            formData.append(`feed[${key}]`, feed[key]);
+        }
+        formData.delete('feed[media]')
+        media.forEach((item, index) => (
+            formData.append('feed[media][]', item)
+        ))
+
+        const url = `http://127.0.0.1:3000/api/v1/feeds/${id}`
```
Comment (haseebuchiha) [inline src/components/Feed/Edit.jsx]:
```suggestion
        const url = `${import.meta.env.BASE_URL}/api/v1/feeds/${id}`
```
Move localhost:3000 address to .env

File: src/components/Feed/Edit.jsx [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+                navigate(`/feeds/${resp.data.id}`)
+            })
+            .catch(resp => { console.log(resp) })
+        setLoading(false)
+    }
+
+    const handleMediaDelete = (id) => {
+        axios.delete(`http://127.0.0.1:3000/api/v1/feed_media/${id}/purge_later`)
+            .then(resp => {
+                console.log(resp)
```
Comment (haseebuchiha) [inline src/components/Feed/Edit.jsx]: remove log

File: src/components/Feed/Edit.jsx [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+                const newMedia = feed.media.filter(item => item.blob_id != resp.data.id)
+                setFeed({
+                    ...feed,
+                    media: newMedia
+                })
+            })
+            .catch(resp => { console.log(resp) })
```
Comment (haseebuchiha) [inline src/components/Feed/Edit.jsx]: remove log

File: src/components/Feed/Feed.jsx [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    const imgFormats = ['jpg', 'jpeg', 'png', 'webp']
+    const vidFormats = ['mp4', 'mov', 'mpeg', 'm4v', 'webm']
+
+    useEffect(() => {
+        const id = params.id
+        const url = `http://127.0.0.1:3000/api/v1/feeds/${id}`
+        axios.get(url)
+            .then(resp => {
+                setFeed(resp.data)
+                console.log(resp.data)
```
Comment (haseebuchiha) [inline src/components/Feed/Feed.jsx]: remove log

File: src/components/Feed/Feed.jsx [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    const changeMedia = () => {
+        if (!media) {
+            media = document.querySelectorAll(".media")
+        }
+        media.forEach((item, index) => {
+            item.style.left = `${(index - currMed) * 100 + 50}%`
+        })
+    }
+
+    function get_url_extension(url) {
```
Comment (haseebuchiha) [inline src/components/Feed/Feed.jsx]: Keep function definitions same

File: src/components/Feed/Feed.jsx [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+                                <img className='object-contain max-h-full max-w-full w-full' src={item.link} />
+                            </div>
+                        }
+                        else if (vidFormats?.includes(get_url_extension(item.link))) {
+                            return <div key={item.blob_id} className="media flex items-center justify-center w-full h-full -translate-x-1/2 absolute transition-all ease-in-out duration-700" style={{ left: `${(index - currMed) * 100 + 50}%` }}>
+                                <ReactPlayer url={item.link} className='object-contain max-h-full max-w-full w-full' width="80%" style={{ left: `${(index - currMed) * 100 + 50}%` }} controls />
+                            </div>
+                        }
+                    })}
+                </div>}
```
Comment (haseebuchiha) [inline src/components/Feed/Feed.jsx]:
```suggestion
                        imgFormats?.includes(get_url_extension(item.link))) ?
                            <div key={item.blob_id} className="media flex items-center justify-center w-full h-full -translate-x-1/2 absolute transition-all ease-in-out duration-700" style={{ left: `${(index - currMed) * 100 + 50}%` }}>
                                <img className='object-contain max-h-full max-w-full w-full' src={item.link} />
                            </div>
                        }
                        : (vidFormats?.includes(get_url_extension(item.link))) {
                            return <div key={item.blob_id} className="media flex items-center justify-center w-full h-full -translate-x-1/2 absolute transition-all ease-in-out duration-700" style={{ left: `${(index - currMed) * 100 + 50}%` }}>
                                <ReactPlayer url={item.link} className='object-contain max-h-full max-w-full w-full' width="80%" style={{ left: `${(index - currMed) * 100 + 50}%` }} controls />
                            </div>
                        }
                    })}
                </div>}
```

partial comment

File: src/components/Feed/Feed.jsx [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+
+
+
+
+
+
+
+
+
+
```
Comment (haseebuchiha) [inline src/components/Feed/Feed.jsx]: why so many empty spaces

File: src/components/Feed/Form.jsx [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    const prevMedia = () => {
+        currMed = (currMed === 0 ? (props.feed.media.length - 1) : currMed - 1)
+        changeMedia()
+    }
+
+    const changeMedia = () => {
+        if (!media) {
+            media = document.querySelectorAll(".media")
```
Comment (haseebuchiha) [inline src/components/Feed/Form.jsx]: useRef

File: src/components/Feed/Form.jsx [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+        return url.split(/[#?]/)[0].split('.').pop().trim();
+    }
+
+    function confirmMediaDelete(id, index) {
+        if (confirm("The media will be deleted even if you do not save the feed. Are you sure?")) {
+            console.log(id)
+            props.handleMediaDelete(id)
+        }
+    }
+
```
Comment (haseebuchiha) [inline src/components/Feed/Form.jsx]: use consistent function naming

File: src/components/Feed/Form.jsx [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+            <div className="mb-4 mt-2 flex align-top w-full">
+                <label className="text-zinc-900 font-semibold inline-block w-48">Description:</label>
+                <textarea onChange={props.handleChange} rows="5" minLength="10" maxLength="1024" className="shadow rounded p-2 text-zinc-900 resize-none w-full" value={props.feed.description} name="description" placeholder="Write description here" />
+            </div>
+
+            {props.feed.media?.length > 0
+                && <div className='relative w-full h-[30rem] mx-8 overflow-hidden cursor-default my-2'>
+                    <ChevronLeftIcon className='h-12 w-12 absolute z-10 top-1/2 -translate-y-1/2 left-4 hover:bg-zinc-400/50 rounded-full transition duration-300 cursor-pointer' onClick={prevMedia} />
+                    <ChevronRightIcon className='h-12 w-12 absolute z-10 top-1/2 -translate-y-1/2 right-4 hover:bg-zinc-400/50 rounded-full transition duration-300 cursor-pointer' onClick={nextMedia} />
+                    {props.feed.media.map((item, index) => {
```
Comment (haseebuchiha) [inline src/components/Feed/Form.jsx:53]: make use of turnary operators

File: src/components/Feed/New.jsx [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+
+const NewFeed = () => {
+    const [feed, setFeed] = useState({})
+    const [media, setMedia] = useState([])
+    const navigate = useNavigate()
+    const [loading, setLoading] = useState(false)
+
+    const handleChange = (e) => {
+        e.preventDefault()
+        setFeed(Object.assign({}, feed, { [e.target.name]: e.target.value }))
```
Comment (haseebuchiha) [inline src/components/Feed/New.jsx]:
```suggestion
        setFeed( {...feed, [e.target.name]: e.target.value})
```

File: src/components/Feed/New.jsx [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+        }
+        media.forEach((item, index) => (
+            formData.append('feed[media][]', item)
+        ))
+
+        axios.post('http://127.0.0.1:3000/api/v1/feeds', formData, { headers: { "Content-Type": "multipart/form-data", } })
+            .then(resp => {
+                navigate(`/feeds/${resp.data.id}`)
+            })
+            .catch(resp => { console.log(resp) })
```
Comment (haseebuchiha) [inline src/components/Feed/New.jsx]: log

File: src/components/Feeds/Feed.jsx [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+            media = document.querySelectorAll(`.media${props.index}`)
+        }
+        media.forEach((item, index) => {
+            item.style.left = `${(index - currMed) * 100 + 50}%`
+        })
+    }
+
+    function get_url_extension(url) {
+        return url.split(/[#?]/)[0].split('.').pop().trim();
+    }
```
Comment (haseebuchiha) [inline src/components/Feeds/Feed.jsx]: Use consistent function naming

File: src/components/Feeds/Feed.jsx [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+            <div className="text-center px-8">
+                <Link to={`/feeds/${props.attributes.id}`} className="text-2xl hover:underline my-2 inline-block">{props.attributes.title}</Link>
+                <hr className="border-0 h-px bg-transparent bg-gradient-to-r from-transparent via-zinc-500 to-transparent opacity-25" />
+                <p className="text-zinc-900 text-start my-4 line-clamp-2">{props.attributes.description}</p>
+                <hr className="border-0 h-px bg-transparent bg-gradient-to-r from-transparent via-zinc-500 to-transparent opacity-25" />
+            </div>
+            {props.attributes.media?.length > 0 && <div className='relative w-full h-[30rem] overflow-hidden cursor-default my-2'>
+                <ChevronLeftIcon className='h-12 w-12 absolute z-10 top-1/2 -translate-y-1/2 left-4 hover:bg-zinc-400/50 rounded-full transition duration-300 cursor-pointer' onClick={prevMedia} />
+                <ChevronRightIcon className='h-12 w-12 absolute z-10 top-1/2 -translate-y-1/2 right-4  hover:bg-zinc-400/50 rounded-full transition duration-300 cursor-pointer' onClick={nextMedia} />
+                {props.attributes.media.map((item, index) => {
```
Comment (haseebuchiha) [inline src/components/Feeds/Feed.jsx:49]: turnary

File: src/components/Feeds/Feeds.jsx [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+import { Link } from 'react-router-dom'
+
+const Feeds = () => {
+    const [feedsArr, setFeedsArr] = useState([])
+    useEffect(() => {
+
+        axios.get('http://127.0.0.1:3000/api/v1/feeds.json')
+            .then(resp => {
+                setFeedsArr(resp.data)
+                console.log(resp.data)
```
Comment (haseebuchiha) [inline src/components/Feeds/Feeds.jsx]: log

File: src/components/Feeds/Feeds.jsx [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+
+    const grid = feedsArr.map((item, index) => {
+        return (
+            <Feed
+                key={item.id}
+                attributes={item}
+                index={index}
+                handleDestroy={handleDestroy}
+            />)
+    })
```
Comment (haseebuchiha) [inline src/components/Feeds/Feeds.jsx]:
```suggestion
    const grid = feedsArr.map((item, index) => 
        (
            <Feed
                key={item.id}
                attributes={item}
                index={index}
                handleDestroy={handleDestroy}
            />)
    )
```

File: src/components/Navbar.jsx [hunk 17]
```diff
@@ -1,56 +1,215 @@
...
+                  <img
+                    className="h-8 w-auto"
+                    src="/vite.svg"
+                    alt="Deltabook"
+                  />
+                </div>
+                <div className="hidden sm:ml-6 sm:block">
+                  <div className="flex space-x-4">
+                    {navigation.map((item) => {
+                      if (item.hasOwnProperty('subitems'))
```
Comment (haseebuchiha) [inline src/components/Navbar.jsx]: tunary

**Review by haseebuchiha — CHANGES_REQUESTED** [review]

**Comment by SaboohTaha** [comment]
Kindly review my code now after I have added requested changes

**Review by haseebuchiha — COMMENTED** [review]
Please resolve conflicts in code.

**Review by haseebuchiha — APPROVED** [review]

**Pull request merged**